File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

def bfs(visit, maps):
	global dx, dy
	que = [(0, 0, 1)]
	while (len(que) > 0):
		x, y, dist = que.pop(0)
		for i in range(4):
			nx = dx[i] + x
			ny = dy[i] + y
			if nx >= 0 and nx < len(maps) and ny >= 0 and ny < len(maps[0]) \
				and maps[nx][ny] == 1 and (visit[nx][ny] == 0 or visit[nx][ny] < dist):
				if visit[nx][ny] == 0 or visit[nx][ny] >= dist:
					visit[nx][ny] = dist + 1
				else:
					continue
				que.append((nx, ny, dist + 1))
	return 

	

def solution(maps):
	answer = 0
	visit = [[0 for _ in range(len(maps[0]))] for _ in range(len(maps))]

	que = [(0, 0, 1)]
	visit[0][0] = 1
	bfs(visit, maps)

	if visit[len(maps) - 1][len(maps[0]) -1] == 0:
		return -1
	return visit[len(maps) - 1][len(maps[0]) -1]

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    int[] dx = new int[]{1, -1, 0, 0};
    int[] dy = new int[]{0, 0, 1, -1};
    public int solution(int[][] maps) {
        int answer = 0;
        Queue<int[]> que = new LinkedList();
        que.add(new int[]{0,0});
        boolean[][] visited = new boolean[maps.length][maps[0].length];
        visited[0][0] = true;
        while (!que.isEmpty()) {
            int[] cur = que.poll();
            for(int i =0 ; i < 4;i++) {
                int nx = dx[i] + cur[0];
                int ny = dy[i] + cur[1];
                if (nx >= 0 && nx < maps.length && ny >= 0 && ny < maps[0].length && !visited[nx][ny] && maps[nx][ny] == 1) {
                    visited[nx][ny] = true;
                    maps[nx][ny] = maps[cur[0]][cur[1]] + 1;
                    que.add(new int[]{nx, ny});
                }
            }
        }
        if (maps[maps.length - 1][maps[0].length - 1] == 1) {
            return -1;
        }
        return maps[maps.length - 1][maps[0].length - 1];
    }
}